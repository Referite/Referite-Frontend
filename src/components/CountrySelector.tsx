import { useEffect, useRef, useState } from 'react';
import '../styles/CountrySelector.css';

interface CountrySelectorProps {
    countriesLst: string[];
    onCountrySelect: (country: string, countryText: string) => any; // Added countryText
    id: string; 
    selectedCountry: any;
    isNewRowAdded: any;
}


export default function CountrySelector ({ countriesLst, onCountrySelect, id, selectedCountry }: CountrySelectorProps)
{

    const [showMenu, setShowMenu] = useState(false);
    const [selectedValue, setSelectedValue] = useState(null);
    const [disabled, ] = useState(false);
    const [disabledItems, ] = useState(new Set<string>());
    const [searchValue, setSearchValue] = useState("");
    const searchRef = useRef<HTMLInputElement>(null);
    const inputRef = useRef<HTMLInputElement>(null);

    const handleItemClick = (option: any) => {
        if (!disabledItems.has(option)) {
            setSelectedValue(option);
            onCountrySelect(option, id);
            setShowMenu(false);

        }
    }

    useEffect(() => {
        const handler = (e: any) => {
            // Close the menu if a click happens outside the dropdown
            if (inputRef.current && !inputRef.current.contains(e.target)) {
                setShowMenu(false);
            }
        };

        window.addEventListener("click", handler);
        return () => {
            window.removeEventListener("click", handler);
        };
    }, []);

    useEffect(() => {
        // Only update the internal state if the prop has actually changed
        if (selectedCountry !== selectedValue) {
          setSelectedValue(selectedCountry || null);
        }
      }, [selectedCountry]); // Only listen to changes of selectedCountry

    const handleInputClick = () => {
        // e.stopPropagation();
        if (!disabled) {
            setShowMenu(!showMenu);
        }
    }

    const isSelected = (option: any) => {
        if (!selectedValue) {
            return false;
        }
        return selectedValue === option;
    }

    const getDisplay = () => {
        return selectedCountry ? <span className="selected">{selectedCountry}</span> : <span className="default">Select Country</span>;
    }

    

    useEffect(() => {
        setSearchValue("");
        if (showMenu && searchRef.current) {
            searchRef.current.focus();
        }
    }, [showMenu]);

    const onSearch = (e: any) => {
        setSearchValue(e.target.value);
    };
    const getOptions = () => {
        if (!searchValue) {
            return countriesLst;
        }
        return countriesLst.filter((country) => country.toLowerCase().indexOf(searchValue.toLowerCase()) >= 0);
    };

    

    return (
        <>
            <div className="input-container-item">
                <div className="dropdown">
                    <div className="country-selector" ref={inputRef} onClick={handleInputClick}> 
                    <div className="country-text" id={id}>
                        {getDisplay()}
                    </div>
                        <div className="dropdown-tools">
                            <div className="dropdown-tool"> âŒ„ </div>
                        </div>
                    </div>
                    {showMenu && (
                        <div className="dropdown-menu">
                        <div className="search-box">
                            <input onChange={onSearch} value={searchValue} ref={searchRef} />
                        </div>
                        {getOptions().map((country) => (
                            <div 
                                onClick={() => handleItemClick(country)}
                                key={country}
                                className={`dropdown-item ${isSelected(country) && "selected"} ${disabledItems.has(country) ? 'disabled' : ''}`}
                            > 
                                {country}
                            </div>
                        ))}
                    </div>
                    )}
                </div>
            </div>  
        </>
    ) 
}
