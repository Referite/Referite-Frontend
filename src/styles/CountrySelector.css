.input-container-item{
    width: 20vw;
    padding-top: 1vw;
    padding-bottom: 1vw;
    font-size: 30px;
    text-align: center;
}

.dropdown {
    grid-area:1/1;
    position: relative;
}

.country-selector {
    display: flex;
    width: 20vw;
    padding-left:1vw;
    border-radius:17px;
    background-color: #f3f3f3;
    box-sizing: border-box;
    border: none;
    height: 3vw;
    word-break: break-all;
    font-size: 1.5vw;
    font-family: 'TH SarabunPSK';
    color: #b9b9b9;
    user-select: none;
    justify-content: space-between;
}

.country-selector:hover {
    cursor: pointer;
}

.country-text {
    padding-top: 3.5%;
}

.dropdown-tools {
    padding-top: 5px;
    font-weight: bold;
    text-align: right;
    margin-right: 5%;
}

.dropdown-menu {
    position: absolute;
    text-align: left;
    font-size: 1.5vw;
    font-family: 'TH SarabunPSK';
    box-sizing: border-box;
    border-radius:8px;
    filter: drop-shadow(2px 2px #c1c1c0);
    background-color: #ffffff;
    width: 100%;
    z-index: 1;
    max-height: 6vw;
    overflow-y: auto;
    transform: translateY(4px);
}

.dropdown-item {
    width: 100%;
}

.dropdown-item:hover {
    background-color: #e8e9f4;
    cursor: pointer;
}

.dropdown-item.selected {
    background-color: #0d6efd;
    color: #fff;
}

.search-box {
    padding: 5px;
    background-color: #eee;
}
.search-box input {
    width: 100%;
    box-sizing: border-box;
    padding: 5px;
    border: 1px solid #ccc;
    border-radius: 5px;
}

.selected {
    color: #5d5b58; 
    font-weight: bold;
}

.default {
    color: #b6b6b6; 
}