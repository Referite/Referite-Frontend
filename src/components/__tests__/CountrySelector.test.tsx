import renderer from 'react-test-renderer';
import CountrySelector from '../CountrySelector';

jest.mock('../CountrySelector', () => 'CountrySelector');
describe('CountrySelector Component', () => {
    it('renders correctly', () => {
      const countriesLst = ['Country1', 'Country2'];
      const onCountrySelect = jest.fn();
  
      const selectedCountry = 'Country1';
  
      const component = renderer.create(
        <CountrySelector
          countriesLst={countriesLst}
          onCountrySelect={onCountrySelect}
          selectedCountry={selectedCountry}
        />
      );
  
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });