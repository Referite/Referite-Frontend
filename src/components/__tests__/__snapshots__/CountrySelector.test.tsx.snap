// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector Component renders correctly 1`] = `
<CountrySelector
  countriesLst={
    [
      "Country1",
      "Country2",
    ]
  }
  onCountrySelect={[MockFunction]}
  selectedCountry="Country1"
/>
`;
